[artifact] only inspect artifact contents

chrisroberts opened the pull request:
### Description

Artifact inspection would inspect the alloc directory for symlinks
pointing outside of the alloc directory. This can be problematic
for drivers like the raw exec driver which provide the root system
and will likely include symlinks.

Inspecting only the artifact after being fetched can be difficult if
the artifact is fetched to a pre-existing destination as it would
require discerning between new and old paths. Instead, if an artifact
is to be inspected, it is fetched to a temporary location. The isolated
artifact is then inspected before being moved to its final destination.

The behavior of the artifact getter will be inconsistent when artifact
inspection is enabled versus when it is disabled. This is due to the 
behavior difference of go-getter with different sources. Because
go-getter does not behave consistently across different source types,
there will be situations where fetching errors may be encountered
when inspection is disabled but fetching is successful when enabled.

### Testing & Reproduction steps

Reproduction steps are described in #26865

### Links

Fixes #26865
<!--
Please include links to GitHub issues, documentation, or similar which is relevant to this PR. If
this is a bug fix, please ensure related issues are linked so they will close when this PR is
merged.
-->

### Contributor Checklist
- [ ] **Changelog Entry** If this PR changes user-facing behavior, please generate and add a
  changelog entry using the `make cl` command.
- [ ] **Testing** Please add tests to cover any new functionality or to demonstrate bug fixes and
  ensure regressions will be caught.
- [ ] **Documentation** If the change impacts user-facing functionality such as the CLI, API, UI,
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
- [ ] **Backport Labels** Please add the correct backport labels as described by the internal
  backporting document.
- [ ] **Commit Type** Ensure the correct merge method is selected which should be "squash and merge"
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
- [ ] **Enterprise PRs** If this is an enterprise only PR, please add any required changelog entry
  within the public repository. 


<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

- [ ] If a change needs to be reverted, we will roll out an update to the code within 7 days.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

tgross commented on client/allocrunner/taskrunner/getter/util.go:
```diff
@@ -189,44 +214,96 @@ func (s *Sandbox) runCmd(env *parameters) error {
+		}
+	}
+
+	if err := mergeDirectories(env.Destination, finalDest); err != nil {
```
If I'm understanding the PR description correctly, this is going to change the behavior of those sources that use option 2, isn't it?

The [artifact.destination](https://developer.hashicorp.com/nomad/docs/job-specification/artifact#destination) is placed relative to the task _working directory_. So couldn't we potentially solve this by having the "root" of the check be the task working directory instead of the allocdir?

chrisroberts replied:
Yes, that is correct that it will change the behavior of those sources using option 2, but only when artifact inspection is triggered. On platforms where landlock is available and is not disabled, artifact inspection will not be performed.

Isolating the check to the task directory won't be sufficient when using the exec driver, as an example. In that case the task directory will include things from the host (like `bin/`) which will contain symlinks.

chrisroberts commented:
Getting back to this: looking at documenting the discrepancies in behavior from using go-getter directly vs redirecting to a temporary directory and then merging in the contents resulted in a large number of differences compounded by edge cases (like a relative destination already existing as a file). Instead, I updated how the temporary destination is setup so that it better mimics the real destination. This allows actual behavior of go-getter to be properly surfaced and stay consistent while still utilizing a temporary directory to allow for inspection. A temporary file within the temporary directory is used to determine if go-getter removed the directory prior to fetch, which allows providing the same behavior.

Added tests are not exhaustive of all the various supported sources and permutations, rather it is demonstrative of the various edge cases to be encountered and shows the same behavior between direct go-getter usage and inspection usage where a temporary directory is first utilized.

tgross commented on client/allocrunner/taskrunner/getter/sandbox_test.go:
```diff
@@ -119,13 +123,20 @@ func TestSandbox_Get_inspection(t *testing.T) {
-	tdir, err := os.MkdirTemp("", "nomad-test")
-	must.NoError(t, err, must.Sprint("failed to create top level local repo directory"))
+	sandboxSetup := func() (string, *Sandbox, interfaces.EnvReplacer) {
+		testutil.RequireRoot(t)
```
~If we've disabled filesystem isolation, shouldn't we be able to avoid requiring root?~ Oh we need it for `os.Chown`. Can we leave a comment about that? It'll make it easier to audit tests that need root in the future.
